x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/BinaryDateTimeFunction.java

```diff
@@ -0,0 +1,131 @@
+
+public abstract class BinaryDateTimeFunction extends BinaryScalarFunction {
+
+    private final ZoneId zoneId;
+    private final BinaryDateOperation operation;
+
+    public BinaryDateTimeFunction(Source source, Expression truncateTo, Expression timestamp, ZoneId zoneId,
+                                  BinaryDateOperation operation) {
+        super(source, truncateTo, timestamp);
+        this.zoneId = zoneId;
+        this.operation = operation;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isString(left(), sourceText(), Expressions.ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (left().foldable()) {
+            String truncateToValue = (String) left().fold();
+            if (truncateToValue != null && resolveDateTimeField(truncateToValue) == false) {
+                List<String> similar = findSimilarDateTimeFields(truncateToValue);
+                if (similar.isEmpty()) {
+                    return new TypeResolution(format(null, "first argument of [{}] must be one of {} or their aliases, found value [{}]",
+                        sourceText(),
+                        validDateTimeFieldValues(),
+                        Expressions.name(left())));
+                } else {
+                    return new TypeResolution(format(null, "Unknown value [{}] for first argument of [{}]; did you mean {}?",
+                        Expressions.name(left()),
+                        sourceText(),
+                        similar));
+                }
+            }
+        }
+        resolution = isDate(right(), sourceText(), Expressions.ParamOrdinal.SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public ZoneId zoneId() {
+        return zoneId;
+    }
+
+    protected abstract boolean resolveDateTimeField(String dateTimeField);
+
+    protected abstract List<String> findSimilarDateTimeFields(String dateTimeField);
+
+    protected abstract List<String> validDateTimeFieldValues();
+
+    @Override
+    protected Pipe makePipe() {
+        return new BinaryDateTimePipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), zoneId, operation);
+    }
+
+    @Override
+    public Nullability nullable() {
+        return Nullability.TRUE;
+    }
+
+    @Override
+    public Object fold() {
+        if (operation == TRUNC) {
+            return DateTruncProcessor.process(left().fold(), right().fold(), zoneId);
+        } else {
+            return DatePartProcessor.process(left().fold(), right().fold(), zoneId);
+        }
+    }
+
+    @Override
+    protected ScriptTemplate asScriptFrom(ScriptTemplate leftScript, ScriptTemplate rightScript) {
+        return new ScriptTemplate(
+            formatTemplate("{sql}." + (operation == TRUNC ? "dateTrunc" : "datePart") +
```

Comment:
Maybe refactor `operation == TRUNC` or, preferable, the entire TRUNC/PART logic in the subclasses themselves so that the abstract class only uses abstract methods that have definitions in the subclasses.

For example, in the `asScriptFrom` method don't use the `operation == TRUNC ? "dateTrunc" : "datePart"` logic,  but delegate the decision to use either `dateTrunc` or `datePart` string to an abstract method belonging to both DatePart and DateTrunc classes where they return `datePart` and `dateTrunc` strings.

And I would say the same should be valid for `fold()` method where the same comparison is being made.